### ISSUE TITLE
Slow nextcloud 15 web login for one particular user

### ISSUE THREAD

**Issue by rainer042** [nextcloud/server#14819]
Hello,

since a while, unfortuantely I cannot tell exactly since when my own user account for my local  nextcloud 15.0.5. When i try to login on my browser it takes about 40 to 90 sec until I see my directory. Another user on the same server just needs perhaps 5 sec to do this. The whlone server is running on Ubuntu 18.04 in a VM on a QNAP Nas TS251. So there is no strong hardware behind the server but it works just fine except for the login of my own user account.

I already deactivated the brute_force app and deleted the whole oc_brute_force_attempts table which had entries before, but this did not help.

What I see when I login in phpmyadmin is that during the time I have to wait a lot of SQL request like 

UPDATE `oc_authtoken` SET `password` = 'P7EAgv2TNrT3VyiBDWpyA/tLZk1MkR5`

are performed which does not happen when I login using the second user whow can login quickly.

my config.php:
<details>

```
<?php
$CONFIG = array (
  'instanceid' => 'octpturgf05m',
  'passwordsalt' => 'xyz',
  'secret' => 'secret',
  'trusted_domains' =>
  array (
    0 => 'owncloud',
    1 => 'owncloud.fritz.box',
    2 => '192.168.11.99',
    3 => 'my.dyndns.de',
  ),
  'memcache.locking' => '\\OC\\Memcache\\Redis',
  'filelocking.enabled' => true,
  'redis' =>
  array (
    'host' => 'localhost',
    'port' => 6379,
  ),
  'auth.bruteforce.protection.enabled' => false,
  'datadirectory' => '/var/www/owncloud/data',
  'overwrite.cli.url' => 'http://owncloud/owncloud',
  'dbtype' => 'mysql',
  'version' => '15.0.5.3',
  'dbname' => 'owncloud',
  'dbhost' => 'localhost',
  'dbtableprefix' => 'oc_',
  'dbuser' => 'oc_admin3',
  'dbpassword' => 'dbpassword',
  'logtimezone' => 'Europe/Berlin',
  'logfile' => '/var/log/owncloud.log',
  'loglevel' => 0,
  'log_authfailip' => true,
  'installed' => true,
  'mail_from_address' => 'ocadmin',
  'mail_smtpmode' => 'smtp',
  'mail_domain' => 'krienke-online.de',
  'mail_smtpsecure' => 'ssl',
  'appstore.experimental.enabled' => true,
  'maintenance' => false,
  'theme' => '',
  'updatechecker' => false,
  'updater.release.channel' => 'stable',
  'updater.secret' => 'anothersecret',
  'mail_smtphost' => 'smtp.provider',
  'mail_smtpauthtype' => 'PLAIN',
  'mail_smtpauth' => 1,
  'mail_smtpname' => 'mymailaccount',
  'mail_smtppassword' => 'password',
  'mail_smtpport' => '465',
);
```
</details>

**Comment by MorrisJobke** [comment]
cc @rullzer @ChristophWurst

**Comment by filthycoding** [comment]
I currently have the same issue, one user that wasn't used for a long time has to wait several minutes for any action to happen, while for any other user, it's quite snappy.

Current workaround is to create a new user and move all data over...

**Comment by rainer042** [comment]
I think its not really useful to simply create a new user and move all data because why should this new user not suffer from the same problem after a while?

I looked closer at my nextcloud and as I said during login of my account that takes such a long time i saw many, many SQLs beeing run like 

UPDATE oc_authtoken SET password = 'P7EAgv2TNrT3VyiBDWpyA/tLZk1MkR5`

This did not happen with another user that can login quickly. So in the web interface of my slow account I found under Settings->Security there in the section "Devices and Sessions" a very long list of devices having access to my account sorted by last access from 1min to 30days. Such an entry contains a small txt depending on the sys client like:

  | Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Thunderbird/60.6.1 Lightning/6.2.6.1 |  
-- | -- | --

This makes sense since I sync calenders and addressbooks stored in nextcloud with several devices (thunderbird, Androis devices etc....) 

Next I manually revoked access to all those entries that were labeled "before 30days" which were the oldest ones I had. I simply clicked on the 3dot menu of an entry and said "revoke" and this about a 100 times for each of "before 30days" entries.  After a logout and a new login with my slow account,  login suddenly was quick again. 
So it seems that at least for me these entries are causing trouble. Does anyone know which database tables are affected by removing one entry so one could automatically remove the oldest entries and thus hopefully get around the slow login problem?

**Comment by ChristophWurst** [comment]
Thank you for the feedback @rainer042. This explains a lot.

So the real issue here is that all these device tokens shouldn't even have been created. Do you use an app password for Thunderbird?

**Comment by rainer042** [comment]
No I didn't use an app password, I simply use my regular login credentials also for calendar and addressbook syncing. 

The strange thing I do not understand why the list of devices and sessions has become that long. I have three or four android devices and also sync with  two different thunderbird instances. I also sometimes use a android filemanager to access nextcloud from different android devices.
But in the list of devices & sessions I see many eg "Sync-client-Linux" entries each with a history from "now" to "30days".  Below such a sync-client-linux there are in turn many entries like I posted above "Mozilla ..." (thunderbird) entries as well as "CaldDav..." entries coming from android devices.  

What I would like to understand is when or why a new entry is created by nextcloud which would perhaps help to understand why so many entries are in my "Devices& Sessions" list  in the end causing the delayed login?

**Comment by ChristophWurst** [comment]
There is logic in place to detect if a client supports sessions. Client connections that use a password but also support sessions will trigger the creation of session tokens. Anything that uses app passwords won't.

So, for some reason, either the server wrongly detects these clients to support sessions (cookies) or the sessions/cookie handling has issues on your setup.

**Comment by rainer042** [comment]
No idea at the moment what might be wrong with my setup in regards to sessions. If the slow login should come back again I will probably create app passwords for my devices accessing nextcloud and if I understood you correctly then no more entries in the sessions list should show up.

**Comment by ChristophWurst** [comment]
Yes, using app passwords should totally fix this.

**Comment by rainer042** [comment]
I found issue  #6203 where you also helped out of quite the same problem. I followed the advice to use unix cron to run cron.php instead of the default Ajax based method. After the first run of /var/www/nextcloud/cron.php via cron all sessions older than 2 days were deleted, so cron.php was probably not active often enough when using Ajax. So I guess using unix cron my problem should be solved as well.
